[WIP] LM Workload

rka97 opened the pull request:
This is for the LM workload.

priyakasimbeg committed:
Merge pull request #847 from mlcommons/dev

Dev -> main

Niccolo-Ajroldi committed:
first LM commit

Niccolo-Ajroldi committed:
lm data pipeline

Niccolo-Ajroldi committed:
testing

Niccolo-Ajroldi committed:
LM workload tested torch pipeline

Niccolo-Ajroldi committed:
LM workload - fix torch tests

Niccolo-Ajroldi committed:
add LM tests, remove dev files

Niccolo-Ajroldi committed:
add LM tests, remove dev files

Niccolo-Ajroldi committed:
Stop tracking .gitignore

Niccolo-Ajroldi committed:
Remove dev/ from repo, keep locally

Niccolo-Ajroldi committed:
fix comments

Niccolo-Ajroldi committed:
add class specifications

Niccolo-Ajroldi committed:
add workload LM info

Niccolo-Ajroldi committed:
restore data_utils.py tree map

Niccolo-Ajroldi committed:
fixed NFS bug

Niccolo-Ajroldi committed:
train/val split before concat

Niccolo-Ajroldi committed:
renamed datasets to avoid conflict with HF

Niccolo-Ajroldi committed:
Merge remote-tracking branch 'upstream/lm_workload' into lm_workload

Niccolo-Ajroldi committed:
renamed datasets to dataset

Niccolo-Ajroldi committed:
fix style

Niccolo-Ajroldi committed:
fix formatting

Niccolo-Ajroldi committed:
fix style

Niccolo-Ajroldi committed:
fix style

Niccolo-Ajroldi committed:
fix yapf

Niccolo-Ajroldi committed:
fix style

rka97 committed:
HF datasets pipeline

rka97 committed:
Testing with linear model

rka97 committed:
Merge branch 'jit_switch' into lm_workload

rka97 committed:
lm workload with linear model

rka97 committed:
add nanodo model

rka97 committed:
torch model

rka97 committed:
lm workload dataset integration in jax

rka97 committed:
lm workload dataset integration in jax

priyakasimbeg committed:
set package versions for transformers and datasets

priyakasimbeg committed:
use train_test_split method to shuffle and split fineweb-edu dataset

priyakasimbeg committed:
modifications to fwedu datasetup

priyakasimbeg committed:
rename fwedu data dir

priyakasimbeg committed:
fix

priyakasimbeg committed:
add back batch mapping in tokenization for fwedu

priyakasimbeg committed:
debugging

priyakasimbeg committed:
debugging

priyakasimbeg committed:
debugging

priyakasimbeg committed:
use tfds to shuffle and split dataset

rka97 committed:
Merge remote-tracking branch 'origin/dev' into lm_workload

priyakasimbeg committed:
add command for fineweb-edu

priyakasimbeg committed:
fix

priyakasimbeg committed:
update calls to sharing utils

rka97 committed:
Fix torch sharding issue, update input pipeline and workload classes to use int32 for tensor types and add dropout rate parameter

rka97 committed:
test working, lm workload training not working (debugging)

priyakasimbeg committed:
updates to input_pipeline and model spec

priyakasimbeg committed:
add defaults for lm workload

priyakasimbeg committed:
refactor eval pipeline and loss fn for lm

priyakasimbeg committed:
refactor evaluation pipeline for lm

priyakasimbeg committed:
remove temporary flag for hlo dumps

priyakasimbeg committed:
fix in workload target condition check

fsschneider commented:
I might be wrong here, but I just randomly saw that we are using a default MLP expansion factor of $4$ (i.e., https://github.com/mlcommons/algorithmic-efficiency/blob/5c85c7e278ffa540d65b1d49f0bd1d0cad732052/algoperf/workloads/lm/lm_pytorch/workload.py#L40).

However, it seems to me, that we are also using a gated linear unit, which uses an initial expansion that is twice as large (since it is split into the gate and the value), i.e. https://github.com/mlcommons/algorithmic-efficiency/blob/5c85c7e278ffa540d65b1d49f0bd1d0cad732052/algoperf/workloads/lm/lm_pytorch/plainlm_model.py#L28-L30

I believe that the common expansion factor of $4$ is for models that use non-gated activation functions (e.g., original Transformer with ReLU or GPT-2 with GELU). To keep the number of parameters constant, people commonly divide by $2/3$ for gated versions. This keeps the total number of parameters in the MLP the same (some small approximation error due to the non-integer multiplication).

If you agree, we should either change the default to $4/3$, or adjust the computation of the `hidden_dim` in the MLP to account for this scaling.

@rka97 @priyakasimbeg @Niccolo-Ajroldi

Niccolo-Ajroldi commented:
@fsschneider agree, we should definitely adjust the expansion factor as you suggested. This was probably an oversight, as in plainLM we rescale when using GLU.

priyakasimbeg commented:
After discussing with Frank, we've decided to handle the modification to the dense layer dimensions in the MLP block itself that adjusts the dimensions based whether we use the gated unit.

I will update the jax workload. This workload is currently ahead of the pytorch one on PR 891. @Niccolo-Ajroldi  I will add you to that PR so you get a good sense of what needs to be updated.

priyakasimbeg committed:
fix in mlp for glu

rka97 committed:
Fix OOM error in weighted cross entropy calculation

rka97 committed:
fix issue with checkpointing bool

priyakasimbeg committed:
increase buffer size

priyakasimbeg committed:
Merge branch 'lm_workload_priya' of github.com:mlcommons/algorithmic-efficiency into lm_workload_priya

priyakasimbeg committed:
remove _eval_batch from jax workload

priyakasimbeg committed:
add todo for pytorch _eval_batch cleanup

rka97 committed:
Merge pull request #891 from mlcommons/lm_workload_priya

Updates to LM PR

priyakasimbeg committed:
add target setting algorithm for fineweb edu lm workload

priyakasimbeg committed:
update step hint for lm workload

priyakasimbeg committed:
update target

priyakasimbeg committed:
update eval split sizes for lm workload and target setting point

rka97 committed:
Porting workload input pipeline to torch

- Added `limit_tf_threads` parameter to `pytorch_init` to control TensorFlow threading based on workload type. Dataloader was going OOM otherwise.
- Updated input pipeline to support "None" for weights (for memory).
- Modified Transformer model's `forward` method to optionally return loss during training. Should be better to fuse the loss later.
- Adjusted torch LM workload configuration for model dimensions and parameters to match jax.
- Updated transformers version in `pyproject.toml`, older version seems unavailable.

rka97 committed:
Merge branch 'lm_workload' of github.com:mlcommons/algorithmic-efficiency into lm_workload

rka97 committed:
Fix OOM bug in lm eval

rka97 committed:
repeat dataset

priyakasimbeg committed:
label smoothing default fix

priyakasimbeg committed:
finish merge

rka97 committed:
Make sure to take the correct number of batches in lm

rka97 committed:
Merge branch 'lm_workload' of github.com:mlcommons/algorithmic-efficiency into lm_workload

rka97 committed:
Properly handle repetition in LM training and evaluation splits

priyakasimbeg committed:
move eval_batch from shared class to framework specific classes since pytorch calls detatch

priyakasimbeg committed:
finish merge

rka97 committed:
Refactor imports and clean up unused code in LM workload and related modules

rka97 committed:
pass linter checks

priyakasimbeg commented on algoperf/workloads/finewebedu_lm/input_pipeline.py:
```diff
@@ -0,0 +1,148 @@
+      },
+    )
+    ds = ds.take(num_batches) if num_batches is not None else ds
+    ds = ds.repeat()
```
@rka97 I don't think we want a repeat here, unless there is something else I'm missing. For the evaluations (both train_eval and validation splits) we should be careful not to repeat evals on examples. For example if we evaluate on the validation set, that means we should account for the predictions on the example in the validation set exactly once. 
If we try to  take more examples than we have available in the split we should raise a StopIteration error, which will happen naturally when the ds runs out of examples.

rka97 replied:
so in eval_model_on_split, we do pass the number of batches

```python

  def _eval_model_on_split(
    self,
    split: str,
    num_examples: int,
    global_batch_size: int,
    params: spec.ParameterContainer,
    model_state: spec.ModelAuxiliaryState,
    rng: spec.RandomState,
    data_dir: str,
    global_step: int = 0,
  ) -> Dict[str, float]:
    """Run a full evaluation of the model."""
    num_batches = int(math.ceil(num_examples / global_batch_size))

    # Handle edge case where num_batches is 0 (e.g., test split with 0 examples)
    if num_batches == 0:
      return {'loss': 0.0, 'ppl': 1.0}

    if split not in self._eval_iters:
      # These iterators will repeat indefinitely.
      self._eval_iters[split] = self._build_input_queue(
        rng, split, data_dir, global_batch_size, num_batches=num_batches
      )
``` 

The above is called by workload.eval_models:
```python

  def eval_model(
    self,
    global_batch_size: int,
    params: ParameterContainer,
    model_state: ModelAuxiliaryState,
    rng: RandomState,
    data_dir: str,
    imagenet_v2_data_dir: Optional[str],
    global_step: int,
  ) -> Dict[str, float]:
    """Run a full evaluation of the model."""
    logging.info('Evaluating on the training split.')
    train_metrics = self._eval_model_on_split(
      split='eval_train',
      num_examples=self.num_eval_train_examples,
      global_batch_size=global_batch_size,
      params=params,
      model_state=model_state,
      rng=rng,
      data_dir=data_dir,
      global_step=global_step,
    )
    eval_metrics = {'train/' + k: v for k, v in train_metrics.items()}
    # We always require a validation set.
    logging.info('Evaluating on the validation split.')
    validation_metrics = self._eval_model_on_split(
      'validation',
      num_examples=self.num_validation_examples,
      global_batch_size=global_batch_size,
      params=params,
      model_state=model_state,
      rng=rng,
      data_dir=data_dir,
      global_step=global_step,
    )
    for k, v in validation_metrics.items():
      eval_metrics['validation/' + k] = v
    eval_metrics['validation/num_examples'] = self.num_validation_examples
    # Evaluate on the test set. TODO(znado): always eval on the test set.
    try:
      if self.num_test_examples is not None:
        logging.info('Evaluating on the test split.')
        test_metrics = self._eval_model_on_split(
          'test',
          num_examples=self.num_test_examples,
          global_batch_size=global_batch_size,
          params=params,
          model_state=model_state,
          rng=rng,
          data_dir=imagenet_v2_data_dir if imagenet_v2_data_dir else data_dir,
          global_step=global_step,
        )
        for k, v in test_metrics.items():
          eval_metrics['test/' + k] = v
        eval_metrics['test/num_examples'] = self.num_test_examples
    except NotImplementedError:
      pass

``` 
which passes the number of validation examples to the workload. This way: we repeat each example exactly once, since the for loop will terminate after num_batches iterations. This allows us to not have to reset the eval dataset loader every time. This seems to be the approach used in other workloads, e.g. in ImageNet [dataloader](https://github.com/mlcommons/algorithmic-efficiency/blob/91988af436f452021e98a61e5144a89d14418e20/algoperf/workloads/imagenet_resnet/imagenet_pytorch/workload.py#L149)
```python
    dataloader = torch.utils.data.DataLoader(
      dataset,
      batch_size=ds_iter_batch_size,
      shuffle=not USE_PYTORCH_DDP and is_train,
      sampler=sampler,
      num_workers=4 if is_train else self.eval_num_workers,
      pin_memory=True,
      drop_last=is_train,
      persistent_workers=is_train,
    )
    dataloader = data_utils.PrefetchedWrapper(dataloader, DEVICE)
    dataloader = data_utils.cycle(
      dataloader,
      custom_sampler=USE_PYTORCH_DDP,
      use_mixup=use_mixup,
      mixup_alpha=0.2,
    )

    return dataloader
```

priyakasimbeg replied:
OK you are correct. I didn't realize we passed the number of examples when we construct the splits as well. Thanks for your thoroughness!

priyakasimbeg commented on algoperf/workloads/finewebedu_lm/input_pipeline.py:
```diff
@@ -0,0 +1,148 @@
+      },
+    )
+    ds = ds.take(num_batches) if num_batches is not None else ds
+    ds = ds.repeat()
```
Same comment as below. I don't think we want to repeat the examples in the eval_train train split.

rka97 committed:
Refactor loss function in LM workloads to unify label handling and improve clarity

rka97 commented:
Adding some TODOs:

- [X] Add model matching tests (i.e. for the same inputs, outputs should be the same, kinda like [model_match](https://github.com/rka97/model_match))
- [X] Fix initialization to be the same for both PyTorch and JAX (there are some minor differences it seems)
- [ ] Add integration test for the lm workload to github actions

rka97 committed:
Fix init in both models to be the same, add lm model diff test

rka97 committed:
Refactor model configuration classes to make them consistent between JAX and PyTorch, also unify initialization to be the same in both

rka97 committed:
Add query-key normalization to CausalAttn and Attention classes, including learned scaling factor

priyakasimbeg commented on algoperf/workloads/finewebedu_lm/finewebedu_lm_jax/models.py:
```diff
@@ -0,0 +1,397 @@
+"""
+Originally based on code from the NanoDO repository under the Apache 2.0 license:
```
Can we rename this file to models.py to be consistent with the pattern in the other workload definitions.

Niccolo-Ajroldi replied:
Agree.

rka97 replied:
Done.

priyakasimbeg commented on algoperf/workloads/finewebedu_lm/finewebedu_lm_pytorch/models.py:
```diff
@@ -0,0 +1,344 @@
+"""
+Originally based on the plainLM codebase:
```
Can we rename this file to models.py to be consistent with the other workload definitions

Niccolo-Ajroldi replied:
Agree.

rka97 replied:
Done.

priyakasimbeg commented on algoperf/workloads/workloads.py:
```diff
@@ -113,6 +113,7 @@
     'workload_path': 'librispeech_deepspeech/librispeech',
     'workload_class_name': 'LibriSpeechDeepSpeechNormAndSpecAugWorkload',
   },
+  'lm': {'workload_path': 'lm/lm', 'workload_class_name': 'LmWorkload'},
```
Now that we have all the important implementation details figured out should we pick a more descriptive name for the workload? I am thinking perhaps 'fineweb_edu_lm'?

Niccolo-Ajroldi replied:
`fineweb_edu_lm` or `finewebedu_lm` make sense, matching the other workload names.

rka97 replied:
Changed to finewebedu_lm.

priyakasimbeg commented on algorithms/archived_paper_baselines/adamw/pytorch/submission.py:
```diff
@@ -189,6 +189,8 @@ def get_batch_size(workload_name):
   elif workload_name == 'mnist':
     return 16
+  elif workload_name == 'lm':
+    return 4
```
This should work for bsz 64 right?

rka97 replied:
Yes, I just made it smaller because I was debugging with v100s. Changed it back to 64.

priyakasimbeg commented on algorithms/archived_paper_baselines/nesterov/jax/submission.py:
```diff
@@ -292,6 +292,8 @@ def get_batch_size(workload_name):
   elif workload_name == 'cifar':
     return 128
+  elif workload_name == 'lm':
+    return 8
```
Should work for bsz 64 right?

rka97 replied:
Yes, I just made it smaller because I was debugging with v100s. Changed it back to 64.

priyakasimbeg reviewed (changes requested):
Second round of small requested changes.

Perhaps something we should discuss, we need a more descriptive name for the workload. E.g. fineweb_edu_lm. What do you all think? @Niccolo-Ajroldi  @rka97

priyakasimbeg committed:
update target

priyakasimbeg committed:
Merge branch 'lm_workload' of github.com:mlcommons/algorithmic-efficiency into lm_workload

priyakasimbeg committed:
add pytorch nadamw_target_setting

priyakasimbeg committed:
docker updates for a100

priyakasimbeg committed:
update budgets for a100 hardware weightclass

priyakasimbeg committed:
formatting

priyakasimbeg committed:
revert changes to docker build shell script

priyakasimbeg committed:
fix merge conflict

priyakasimbeg committed:
merge

rka97 committed:
rename models.py

rka97 committed:
rename workload

priyakasimbeg committed:
update budget for lm

priyakasimbeg committed:
Merge branch 'lm_workload' of github.com:mlcommons/algorithmic-efficiency into lm_workload

priyakasimbeg committed:
update pytorch

priyakasimbeg committed:
Merge branch 'a100' into lm_workload

rka97 committed:
add mixed precision training

rka97 committed:
ImageNet and CIFAR mixed-precision support, need to debug slow pytorch

- Introduced DTYPE enum to standardize data types (FLOAT32, FLOAT16, BFLOAT16) for JAX and PyTorch.
- Updated input pipelines and model definitions in CIFAR and ImageNet workloads to utilize mixed precision.
- Implemented casting policies for parameters and inputs using jmp and torch.autocast.

priyakasimbeg committed:
adjust step time calculation

priyakasimbeg committed:
Merge branch 'lm_workload' of github.com:mlcommons/algorithmic-efficiency into lm_workload

priyakasimbeg committed:
merge

priyakasimbeg committed:
set target to target found w narrower search space

priyakasimbeg committed:
add finewebedu_lm to docker startup script valid args
